File: Sources/KakaoSapiens/Services/SyncPairingCoordinator.swift
import CryptoKit
import Foundation

/// Joining an existing account, from both ends.
///
/// Two roles that know different things. The host already holds the account
/// master key and decides whether to let a device in; the joiner holds nothing
/// and only builds its own device identity. The Worker between them is a
/// mailbox: the plaintext master key never passes through it, because what the
/// joiner receives is a package the host sealed under a key derived from the
/// `pairing_secret` that was handed over by screen.
public enum SyncPairingError: Error, Equatable {
    case notConnected
    case sessionExpired
    case noSuchClaim
    /// The user has not confirmed the two screens show the same number.
    case shortAuthenticationStringNotConfirmed
    case rejected
    case storageFailed
    case transport
}

/// What the host learned about one waiting device.
public struct SyncPairingCandidate: Equatable {
    public let claimID: String
    public let claimLookup: Data
    /// The six digits the user compares against the joiner's screen.
    public let shortAuthenticationString: String
    let deviceID: String
    let spaceID: String
    let platform: String
    let claimSecret: Data
}

/// The host side: the device that is already in.
public actor SyncPairingHostCoordinator {
    private let client: SyncPairingClient
    private let secrets: () -> SyncSecretLoadResult
    private let random: SyncRandomSource

    private var sessionID: String?
    private var pairingSecret: Data?

    public init(
        client: SyncPairingClient,
        secrets: @escaping () -> SyncSecretLoadResult,
        random: SyncRandomSource
    ) {
        self.client = client
        self.secrets = secrets
        self.random = random
    }

    /// Open a session and produce the payload to show. Only a device that is
    /// actually connected may do this — an unconnected one has no master key
    /// to hand over and would strand whoever scanned it.
    public func openSession(accountID: String, baseURL: URL) async throws -> SyncPairingPayload {
        guard case .available = secrets() else { throw SyncPairingError.notConnected }
        let secret = random.bytes(32)
        let material = try? SyncE2EE.derivePairingMaterial(pairingSecret: secret, claimSecret: secret)
        guard let lookup = material?.pairingSessionLookup else { throw SyncPairingError.storageFailed }
        let session = Self.uuid(random)

        do {
            _ = try await client.createSession(sessionID: session, sessionLookup: lookup)
        } catch let error as SyncPairingClient.PairingClientError {
            throw Self.map(error)
        }

        sessionID = session
        pairingSecret = secret
        return try SyncPairingPayload(
            baseURL: baseURL,
            accountID: accountID,
            sessionID: session,
            pairingSecret: secret
        )
    }

    /// Read waiting claims and open each one. A claim this host cannot open is
    /// not for this session and is dropped rather than shown.
    public func pollCandidates() async throws -> [SyncPairingCandidate] {
        guard let session = sessionID, let secret = pairingSecret else {
            throw SyncPairingError.notConnected
        }
        let claims: [SyncPairingClient.Claim]
        do {
            claims = try await client.listClaims(sessionID: session)
        } catch let error as SyncPairingClient.PairingClientError {
            throw Self.map(error)
        }

        return claims.compactMap { claim -> SyncPairingCandidate? in
            guard claim.state == "submitted",
                  let lookup = Data(base64Encoded: claim.claimLookup),
                  let envelope = Data(base64Encoded: claim.claimEnvelope) else { return nil }
            // The claim key comes from the pairing secret alone, so only a host
            // holding the secret it showed can read what a scanner submitted.
            guard let claimKey = try? SyncE2EE.derivePairingMaterial(
                pairingSecret: secret,
                claimSecret: secret
            ).pairingClaimKey else { return nil }
            guard let plaintext = try? SyncE2EE.openPairing(
                envelope: envelope,
                key: claimKey,
                sessionID: session,
                claimID: claim.claimID,
                claimLookup: lookup,
                payloadType: .claim
            ) else { return nil }
            guard let body = try? JSONSerialization.jsonObject(with: plaintext) as? [String: Any],
                  let deviceID = body["device_id"] as? String,
                  let spaceID = body["space_id"] as? String,
                  let platform = body["platform"] as? String,
                  let claimSecretText = body["claim_secret"] as? String,
                  let claimSecret = Data(base64Encoded: claimSecretText) else { return nil }
            guard let material = try? SyncE2EE.derivePairingMaterial(
                pairingSecret: secret,
                claimSecret: claimSecret
            ), material.claimLookup == lookup else { return nil }

            return SyncPairingCandidate(
                claimID: claim.claimID,
                claimLookup: lookup,
                shortAuthenticationString: material.pairingSAS,
                deviceID: deviceID,
                spaceID: spaceID,
                platform: platform,
                claimSecret: claimSecret
            )
        }
    }

    /// Approve exactly one candidate, after the user says the numbers match.
    ///
    /// `sasConfirmed` is not a formality. Without a human comparing the two
    /// screens, an attacker who relayed the QR could be the one waiting, and
    /// approving would hand them the account master key.
    public func approve(
        _ candidate: SyncPairingCandidate,
        sasConfirmed: Bool
    ) async throws {
        guard sasConfirmed else { throw SyncPairingError.shortAuthenticationStringNotConfirmed }
        guard let session = sessionID, let secret = pairingSecret else {
            throw SyncPairingError.notConnected
        }
        guard case .available(let bundle) = secrets() else { throw SyncPairingError.notConnected }
        guard let material = try? SyncE2EE.derivePairingMaterial(
            pairingSecret: secret,
            claimSecret: candidate.claimSecret
        ) else { throw SyncPairingError.storageFailed }

        // A token the joiner will authenticate with. Only its hash goes to the
        // Worker; the value itself travels inside the sealed package.
        let newToken = random.bytes(32)
        let tokenHash = Data(SHA256.hash(data: newToken)).map { String(format: "%02x", $0) }.joined()
        // No account id here: the joiner already read it from the payload it
        // scanned, and a second copy inside the package would be a second
        // answer to the same question.
        let delivery: [String: Any] = [
            "account_master_key": bundle.accountMasterKey.base64EncodedString(),
            "device_token": newToken.base64EncodedString(),
        ]
        guard let plaintext = try? JSONSerialization.data(withJSONObject: delivery, options: [.sortedKeys]),
              let envelope = try? SyncE2EE.sealPairing(
                  plaintext: plaintext,
                  key: material.pairingDeliveryKey,
                  nonce: random.bytes(12),
                  sessionID: session,
                  claimID: candidate.claimID,
                  claimLookup: candidate.claimLookup,
                  payloadType: .delivery
              ) else { throw SyncPairingError.storageFailed }

        do {
            try await client.approve(
                sessionID: session,
                claimID: candidate.claimID,
                claimLookup: candidate.claimLookup,
                deliveryEnvelope: envelope,
                device: (
                    id: candidate.deviceID,
                    spaceID: candidate.spaceID,
                    platform: candidate.platform,
                    tokenHash: tokenHash
                )
            )
        } catch let error as SyncPairingClient.PairingClientError {
            throw Self.map(error)
        }
    }

    static func uuid(_ random: SyncRandomSource) -> String {
        var bytes = [UInt8](random.bytes(16))
        bytes[6] = (bytes[6] & 0x0f) | 0x40
        bytes[8] = (bytes[8] & 0x3f) | 0x80
        let hex = bytes.map { String(format: "%02X", $0) }.joined()
        let ranges = [0..<8, 8..<12, 12..<16, 16..<20, 20..<32]
        return ranges.map { range in
            String(hex[hex.index(hex.startIndex, offsetBy: range.lowerBound)..<hex.index(hex.startIndex, offsetBy: range.upperBound)])
        }.joined(separator: "-")
    }

    static func map(_ error: SyncPairingClient.PairingClientError) -> SyncPairingError {
        switch error {
        case .httpStatus(let status) where status == 404 || status == 410: return .sessionExpired
        case .httpStatus(let status) where status == 409: return .rejected
        case .httpStatus: return .rejected
        case .notAuthenticated: return .notConnected
        case .malformedResponse: return .transport
        }
    }
}

/// The joiner side: a device that belongs to no account yet.
public actor SyncPairingJoinerCoordinator {
    private let random: SyncRandomSource
    private let vault: SyncSecretVault
    private let connectionStore: SyncConnectionStateStore

    private var payload: SyncPairingPayload?
    private var claimID: String?
    private var claimSecret: Data?
    private var material: SyncE2EE.PairingMaterial?

    public init(
        random: SyncRandomSource,
        vault: SyncSecretVault,
        connectionStore: SyncConnectionStateStore
    ) {
        self.random = random
        self.vault = vault
        self.connectionStore = connectionStore
    }

    /// Accept a scanned payload and build this device's claim. Nothing is sent.
    public func accept(text: String, deviceID: String, spaceID: String, platform: String) throws
        -> (payload: SyncPairingPayload, claimID: String, shortAuthenticationString: String) {
        let scanned = try SyncPairingPayload.decode(text: text)
        let secret = random.bytes(32)
        let claim = SyncPairingHostCoordinator.uuid(random)
        guard let derived = try? SyncE2EE.derivePairingMaterial(
            pairingSecret: scanned.pairingSecret,
            claimSecret: secret
        ) else { throw SyncPairingError.storageFailed }

        payload = scanned
        claimID = claim
        claimSecret = secret
        material = derived
        self.deviceIdentity = (deviceID, spaceID, platform)
        return (scanned, claim, derived.pairingSAS)
    }

    private var deviceIdentity: (id: String, space: String, platform: String)?

    public func submit(using client: SyncPairingClient) async throws {
        guard let scanned = payload, let claim = claimID, let secret = claimSecret,
              let derived = material, let identity = deviceIdentity else {
            throw SyncPairingError.notConnected
        }
        let body: [String: Any] = [
            "device_id": identity.id,
            "space_id": identity.space,
            "platform": identity.platform,
            "claim_secret": secret.base64EncodedString(),
        ]
        guard let plaintext = try? JSONSerialization.data(withJSONObject: body, options: [.sortedKeys]),
              let hostKey = try? SyncE2EE.derivePairingMaterial(
                  pairingSecret: scanned.pairingSecret,
                  claimSecret: scanned.pairingSecret
              ).pairingClaimKey,
              let envelope = try? SyncE2EE.sealPairing(
                  plaintext: plaintext,
                  key: hostKey,
                  nonce: random.bytes(12),
                  sessionID: scanned.sessionID,
                  claimID: claim,
                  claimLookup: derived.claimLookup,
                  payloadType: .claim
              ),
              let verifier = try? SyncE2EE.claimRedeemVerifier(
                  sessionID: scanned.sessionID,
                  claimID: claim,
                  claimLookup: derived.claimLookup,
                  claimRedeemAuth: derived.claimRedeemAuth
              ) else { throw SyncPairingError.storageFailed }

        do {
            try await client.submitClaim(
                sessionID: scanned.sessionID,
                sessionLookup: derived.pairingSessionLookup,
                claimID: claim,
                claimLookup: derived.claimLookup,
                claimEnvelope: envelope,
                redeemVerifier: verifier.map { String(format: "%02x", $0) }.joined()
            )
        } catch let error as SyncPairingClient.PairingClientError {
            throw SyncPairingHostCoordinator.map(error)
        }
    }

    /// Redeem once and store what came back.
    ///
    /// Nothing is written until the package opens and its contents check out.
    /// A half-linked device — secrets without a connection, or the reverse — is
    /// a state the user has no way to repair.
    public func redeem(using client: SyncPairingClient, sasConfirmed: Bool) async throws {
        guard sasConfirmed else { throw SyncPairingError.shortAuthenticationStringNotConfirmed }
        guard let scanned = payload, let claim = claimID, let derived = material,
              let identity = deviceIdentity else {
            throw SyncPairingError.notConnected
        }

        let envelope: Data
        do {
            envelope = try await client.redeem(
                sessionID: scanned.sessionID,
                claimID: claim,
                claimLookup: derived.claimLookup,
                redeemAuth: derived.claimRedeemAuth
            )
        } catch let error as SyncPairingClient.PairingClientError {
            throw SyncPairingHostCoordinator.map(error)
        }

        guard let plaintext = try? SyncE2EE.openPairing(
            envelope: envelope,
            key: derived.pairingDeliveryKey,
            sessionID: scanned.sessionID,
            claimID: claim,
            claimLookup: derived.claimLookup,
            payloadType: .delivery
        ) else { throw SyncPairingError.rejected }

        guard let body = try? JSONSerialization.jsonObject(with: plaintext) as? [String: Any],
              let masterText = body["account_master_key"] as? String,
              let tokenText = body["device_token"] as? String,
              let master = Data(base64Encoded: masterText),
              let token = Data(base64Encoded: tokenText),
              let bundle = try? SyncSecretBundle(accountMasterKey: master, deviceToken: token) else {
            throw SyncPairingError.rejected
        }

        do { try vault.save(bundle) } catch { throw SyncPairingError.storageFailed }
        // The account is the host's, and synchronisation stays off: joining is
        // a connection, not a decision to start syncing.
        guard let configuration = try? SyncConnectionConfiguration(
            baseURL: scanned.baseURL,
            accountID: scanned.accountID,
            deviceID: identity.id,
            enabled: false,
            changesCursor: nil
        ), (try? connectionStore.save(configuration)) != nil else {
            throw SyncPairingError.storageFailed
        }
    }
}

File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncPairingCoordinator.kt
package com.sapiens.gagaodok.sync

import java.security.MessageDigest
import java.util.Base64
import org.json.JSONObject

/**
 * Joining an existing account, from both ends.
 *
 * Two roles that know different things. The host already holds the account
 * master key and decides whether to let a device in; the joiner holds nothing
 * and only builds its own device identity. The Worker between them is a
 * mailbox: the plaintext master key never passes through it, because what the
 * joiner receives is a package the host sealed under a key derived from the
 * `pairing_secret` that was handed over by screen.
 */
class SyncPairingException(val reason: Reason) : Exception() {
    enum class Reason {
        NOT_CONNECTED,
        SESSION_EXPIRED,
        /** The user has not confirmed the two screens show the same number. */
        SAS_NOT_CONFIRMED,
        REJECTED,
        STORAGE_FAILED,
        TRANSPORT,
    }
}

/** What the host learned about one waiting device. */
data class SyncPairingCandidate(
    val claimId: String,
    val claimLookup: ByteArray,
    /** The six digits the user compares against the joiner's screen. */
    val shortAuthenticationString: String,
    internal val deviceId: String,
    internal val spaceId: String,
    internal val platform: String,
    internal val claimSecret: ByteArray,
)

/** The host side: the device that is already in. */
class SyncPairingHostCoordinator(
    private val client: SyncPairingClient,
    private val secrets: () -> SyncSecretLoadResult,
    private val random: SyncRandomSource,
) {
    private var sessionId: String? = null
    private var pairingSecret: ByteArray? = null

    /**
     * Open a session and produce the payload to show. Only a device that is
     * actually connected may do this — an unconnected one has no master key to
     * hand over and would strand whoever scanned it.
     */
    fun openSession(accountId: String, baseUrl: String): SyncPairingPayload {
        if (secrets() !is SyncSecretLoadResult.Available) {
            throw SyncPairingException(SyncPairingException.Reason.NOT_CONNECTED)
        }
        val secret = random.bytes(32)
        val lookup = runCatching {
            SyncE2EE.derivePairingMaterial(secret, secret).pairingSessionLookup
        }.getOrElse { throw SyncPairingException(SyncPairingException.Reason.STORAGE_FAILED) }
        val session = uuid(random)

        runCatching { client.createSession(session, lookup) }
            .getOrElse { throw map(it) }

        sessionId = session
        pairingSecret = secret
        return SyncPairingPayload.create(baseUrl, accountId, session, secret)
    }

    /**
     * Read waiting claims and open each one. A claim this host cannot open is
     * not for this session and is dropped rather than shown.
     */
    fun pollCandidates(): List<SyncPairingCandidate> {
        val session = sessionId
        val secret = pairingSecret
        if (session == null || secret == null) {
            throw SyncPairingException(SyncPairingException.Reason.NOT_CONNECTED)
        }
        val claims = runCatching { client.listClaims(session) }.getOrElse { throw map(it) }

        return claims.mapNotNull { claim ->
            runCatching {
                if (claim.state != "submitted") return@runCatching null
                val lookup = Base64.getDecoder().decode(claim.claimLookup)
                val envelope = Base64.getDecoder().decode(claim.claimEnvelope)
                // The claim key comes from the pairing secret alone, so only a
                // host holding the secret it showed can read a submission.
                val claimKey = SyncE2EE.derivePairingMaterial(secret, secret).pairingClaimKey
                val plaintext = SyncE2EE.openPairing(
                    envelope, claimKey, session, claim.claimId, lookup,
                    SyncE2EE.PairingPayloadType.CLAIM,
                )
                val body = JSONObject(String(plaintext, Charsets.UTF_8))
                val claimSecret = Base64.getDecoder().decode(body.getString("claim_secret"))
                val material = SyncE2EE.derivePairingMaterial(secret, claimSecret)
                if (!material.claimLookup.contentEquals(lookup)) return@runCatching null
                SyncPairingCandidate(
                    claimId = claim.claimId,
                    claimLookup = lookup,
                    shortAuthenticationString = material.pairingSAS,
                    deviceId = body.getString("device_id"),
                    spaceId = body.getString("space_id"),
                    platform = body.getString("platform"),
                    claimSecret = claimSecret,
                )
            }.getOrNull()
        }
    }

    /**
     * Approve exactly one candidate, after the user says the numbers match.
     *
     * `sasConfirmed` is not a formality. Without a human comparing the two
     * screens, an attacker who relayed the QR could be the one waiting, and
     * approving would hand them the account master key.
     */
    fun approve(candidate: SyncPairingCandidate, sasConfirmed: Boolean) {
        if (!sasConfirmed) throw SyncPairingException(SyncPairingException.Reason.SAS_NOT_CONFIRMED)
        val session = sessionId
        val secret = pairingSecret
        if (session == null || secret == null) {
            throw SyncPairingException(SyncPairingException.Reason.NOT_CONNECTED)
        }
        val bundle = (secrets() as? SyncSecretLoadResult.Available)?.secrets
            ?: throw SyncPairingException(SyncPairingException.Reason.NOT_CONNECTED)

        val newToken = random.bytes(32)
        val envelope = runCatching {
            val material = SyncE2EE.derivePairingMaterial(secret, candidate.claimSecret)
            // No account id here: the joiner already read it from the payload it
            // scanned, and a second copy would be a second answer to one question.
            val delivery = JSONObject()
                .put("account_master_key", Base64.getEncoder().encodeToString(bundle.accountMasterKey))
                .put("device_token", Base64.getEncoder().encodeToString(newToken))
            SyncE2EE.sealPairing(
                delivery.toString().toByteArray(Charsets.UTF_8),
                material.pairingDeliveryKey,
                random.bytes(12),
                session,
                candidate.claimId,
                candidate.claimLookup,
                SyncE2EE.PairingPayloadType.DELIVERY,
            )
        }.getOrElse { throw SyncPairingException(SyncPairingException.Reason.STORAGE_FAILED) }

        // Only the hash goes to the Worker; the token itself travels sealed.
        val tokenHash = MessageDigest.getInstance("SHA-256").digest(newToken)
            .joinToString("") { "%02x".format(it) }
        runCatching {
            client.approve(
                sessionId = session,
                claimId = candidate.claimId,
                claimLookup = candidate.claimLookup,
                deliveryEnvelope = envelope,
                deviceId = candidate.deviceId,
                spaceId = candidate.spaceId,
                platform = candidate.platform,
                deviceTokenHash = tokenHash,
            )
        }.getOrElse { throw map(it) }
    }

    internal companion object {
        fun uuid(random: SyncRandomSource): String {
            val bytes = random.bytes(16)
            bytes[6] = ((bytes[6].toInt() and 0x0f) or 0x40).toByte()
            bytes[8] = ((bytes[8].toInt() and 0x3f) or 0x80).toByte()
            val hex = bytes.joinToString("") { "%02X".format(it) }
            return listOf(
                hex.substring(0, 8), hex.substring(8, 12), hex.substring(12, 16),
                hex.substring(16, 20), hex.substring(20, 32),
            ).joinToString("-")
        }

        fun map(error: Throwable): SyncPairingException = when {
            error is SyncPairingException -> error
            error is SyncPairingClient.PairingClientException && error.notAuthenticated ->
                SyncPairingException(SyncPairingException.Reason.NOT_CONNECTED)
            error is SyncPairingClient.PairingClientException &&
                (error.statusCode == 404 || error.statusCode == 410) ->
                SyncPairingException(SyncPairingException.Reason.SESSION_EXPIRED)
            error is SyncPairingClient.PairingClientException && error.statusCode != null ->
                SyncPairingException(SyncPairingException.Reason.REJECTED)
            else -> SyncPairingException(SyncPairingException.Reason.TRANSPORT)
        }
    }
}

data class SyncPairingRedeemedCandidate(
    val secrets: SyncSecretBundle,
    val connection: SyncConnectionConfiguration,
)

/** The joiner side: a device that belongs to no account yet. */
class SyncPairingJoinerCoordinator(
    private val random: SyncRandomSource,
    private val vault: SyncSecretVault,
    private val connectionStore: SyncConnectionStateStore,
) {
    data class Accepted(
        val payload: SyncPairingPayload,
        val claimId: String,
        val shortAuthenticationString: String,
    )

    private var payload: SyncPairingPayload? = null
    private var claimId: String? = null
    private var claimSecret: ByteArray? = null
    private var material: SyncE2EE.PairingMaterial? = null
    private var identity: Triple<String, String, String>? = null

    /** Accept a scanned payload and build this device's claim. Nothing is sent. */
    fun accept(text: String, deviceId: String, spaceId: String, platform: String): Accepted {
        val scanned = SyncPairingPayload.decode(text)
        val secret = random.bytes(32)
        val claim = SyncPairingHostCoordinator.uuid(random)
        val derived = runCatching {
            SyncE2EE.derivePairingMaterial(scanned.pairingSecret, secret)
        }.getOrElse { throw SyncPairingException(SyncPairingException.Reason.STORAGE_FAILED) }

        payload = scanned
        claimId = claim
        claimSecret = secret
        material = derived
        identity = Triple(deviceId, spaceId, platform)
        return Accepted(scanned, claim, derived.pairingSAS)
    }

    fun submit(client: SyncPairingClient) {
        val scanned = payload
        val claim = claimId
        val secret = claimSecret
        val derived = material
        val who = identity
        if (scanned == null || claim == null || secret == null || derived == null || who == null) {
            throw SyncPairingException(SyncPairingException.Reason.NOT_CONNECTED)
        }
        val prepared = runCatching {
            val body = JSONObject()
                .put("device_id", who.first)
                .put("space_id", who.second)
                .put("platform", who.third)
                .put("claim_secret", Base64.getEncoder().encodeToString(secret))
            val hostKey = SyncE2EE.derivePairingMaterial(
                scanned.pairingSecret, scanned.pairingSecret,
            ).pairingClaimKey
            val envelope = SyncE2EE.sealPairing(
                body.toString().toByteArray(Charsets.UTF_8),
                hostKey,
                random.bytes(12),
                scanned.sessionId,
                claim,
                derived.claimLookup,
                SyncE2EE.PairingPayloadType.CLAIM,
            )
            val verifier = SyncE2EE.claimRedeemVerifier(
                scanned.sessionId, claim, derived.claimLookup, derived.claimRedeemAuth,
            ).joinToString("") { "%02x".format(it) }
            envelope to verifier
        }.getOrElse { throw SyncPairingException(SyncPairingException.Reason.STORAGE_FAILED) }

        runCatching {
            client.submitClaim(
                sessionId = scanned.sessionId,
                sessionLookup = derived.pairingSessionLookup,
                claimId = claim,
                claimLookup = derived.claimLookup,
                claimEnvelope = prepared.first,
                redeemVerifier = prepared.second,
            )
        }.getOrElse { throw SyncPairingHostCoordinator.map(it) }
    }

    /**
     * Redeem once and store what came back.
     *
     * Nothing is written until the package opens and its contents check out. A
     * half-linked device — secrets without a connection, or the reverse — is a
     * state the user has no way to repair.
     */
    fun redeemCandidate(client: SyncPairingClient, sasConfirmed: Boolean): SyncPairingRedeemedCandidate {
        if (!sasConfirmed) throw SyncPairingException(SyncPairingException.Reason.SAS_NOT_CONFIRMED)
        val scanned = payload
        val claim = claimId
        val derived = material
        val who = identity
        if (scanned == null || claim == null || derived == null || who == null) {
            throw SyncPairingException(SyncPairingException.Reason.NOT_CONNECTED)
        }

        val envelope = runCatching {
            client.redeem(scanned.sessionId, claim, derived.claimLookup, derived.claimRedeemAuth)
        }.getOrElse { throw SyncPairingHostCoordinator.map(it) }

        val bundle = runCatching {
            val plaintext = SyncE2EE.openPairing(
                envelope, derived.pairingDeliveryKey, scanned.sessionId, claim,
                derived.claimLookup, SyncE2EE.PairingPayloadType.DELIVERY,
            )
            val body = JSONObject(String(plaintext, Charsets.UTF_8))
            SyncSecretBundle(
                Base64.getDecoder().decode(body.getString("account_master_key")),
                Base64.getDecoder().decode(body.getString("device_token")),
            )
        }.getOrElse { throw SyncPairingException(SyncPairingException.Reason.REJECTED) }

        return SyncPairingRedeemedCandidate(
            bundle,
            SyncConnectionConfiguration(
                baseUrl = scanned.baseUrl,
                accountId = scanned.accountId,
                deviceId = who.first,
                enabled = false,
                changesCursor = null,
            ),
        )
    }

    fun redeem(client: SyncPairingClient, sasConfirmed: Boolean) {
        val candidate = redeemCandidate(client, sasConfirmed)
        if (!vault.save(candidate.secrets)) throw SyncPairingException(SyncPairingException.Reason.STORAGE_FAILED)
        // The account is the host's, and synchronisation stays off: joining is
        // a connection, not a decision to start syncing.
        val saved = runCatching { connectionStore.save(candidate.connection) }.getOrDefault(false)
        if (!saved) throw SyncPairingException(SyncPairingException.Reason.STORAGE_FAILED)
    }
}
